Add kiosk and main page

theScrabi opened the pull request:
# Prototype do not merge!!!

This will replage the Blank screen with a Kiosk page. At the moment it's Youtube "Trending".

Things that need to be done:
- [x] Make Items touchable (results in stack trace right now)
- [x] Remove back arrow
- [x] Add option to settings to bring "Black page back"
- [x] add a proper handling for current service in [main_fragment](https://github.com/TeamNewPipe/NewPipe/blob/trending/app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java#L30)
- [x] add proper error handling when creating [instance](https://github.com/TeamNewPipe/NewPipe/blob/trending/app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java#L130) of KioskFragment
- [x] Make type/name/id translatable
- [x] s/kisok/kiosk/g

This is what it looks like atm:
![screenshot_1506181400](https://user-images.githubusercontent.com/1891273/30774673-d04d0be4-a086-11e7-9767-5127465bd228.png)

wb9688 commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -18,11 +18,16 @@
+    private boolean showBlankTab = false;
+
+    //todo: FIX THIS URGENTLY
+    public int currentServiceId = 0; //for youtube
```
If #693 is merged, you could do `Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(getActivity()).getString("service", "0"))`.

wb9688 replied:
Yeah, but #693 isn't merged yet, so you'd have to wait for that.

theScrabi replied:
hm, don't you think it was possible to introduc it anyway, I mean its no code that will colide i guess.

wb9688 replied:
You mean merging #693? Well, that's possible, but then I'd at least add the correct intents for SoundCloud (can you do that? I know nothing about Android's intents xD). After that I can just make another PR for the new service layout.

theScrabi replied:
No don't merge #693 sorry. No I ment adding that line `Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(getActivity()).getString("service", "0")).` wouldn't make any code collide. So we could put it into fragment_main anyway. ... What do you think?

wb9688 replied:
Oh, if that doesn't throw an error, you should add it.

mauriciocolli replied:
It don't, but we are hardcoding keys (see settings_key.xml). Also, why are we parsing a `String` when we could save the `int` directly?

wb9688 replied:
@mauriciocolli: I tried doing it with `int`s, but I got some weird error. If you could get it working, please commit that to #693.

theScrabi replied:
Well we we could put it into settings_keys.xml than its just one line that collides.

wb9688 commented on app/src/main/java/org/schabi/newpipe/fragments/list/kisok/KioskFragment.java:
```diff
@@ -0,0 +1,148 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with OpenHitboxStreams.  If not, see <http://www.gnu.org/licenses/>.
```
I think you meant NewPipe instead of OpenHitboxStreams. ;)

theScrabi replied:
Yea that's copypasta, I took the code from one of my other apps.

mauriciocolli replied:
Also there's a dangling javadoc comment there (use /* instead of /** for copyright notice, since you're not documenting).

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/list/kisok/KioskFragment.java:
```diff
@@ -0,0 +1,148 @@
+        super.handleResult(result);
+
+        animateView(headerRootLayout, true, 100);
+        headerTitleView.setText(result.type);
```
You are right. Id/type/name was not meant to be displayed directly.
I'll add a todo.

wb9688 replied:
Yeah, I was thinking about it, but I couldn't find a better way than bunch of `if` statements, so do you have any better ideas?

theScrabi replied:
I see. Isn't it possible to get some stringresource like R.string.<some_string> programmatically? Something like `Resource.getString("some_string");` If not than we might have to find another solution.

mauriciocolli replied:
I thought in doing a "generify" and use the most common term as the id/type, because it is more convenient to use a generic term to cover a lot of cases. 

_Example:_ SoundCloud uses "New & Hot", YouTube's equivalent is "Trending".
In this case, "Trending" is the selected one (as it is more known).

_Example 2:_ A lot of services use "Top ##", this can easily be "generified" because only the number changes.

###### Though there's a downside to it, we lost the "uniqueness" of each service.

theScrabi replied:
I don't know I don't like generalizing either.
But why don't we just crawl the Name.

mauriciocolli replied:
I thought in generify the extractor and then _**in the NewPipe itself**_, make a unique translation for each service, much better than hardcode-checking in this repository (and translating in the extractor repository would force us to use some way to get the translations).

E.g.:
```java
switch (kiosk.getType()) {
    case Top:
        getString(soundcloud_top); // Top
    case Trending:
        getString(soundcloud_trending); // New & Hot
}
```

>But why don't we just crawl the Name

From where? SoundCloud for example, doesn't have all languages that NewPipe support.

wb9688 replied:
Yeah, I don't like 'generifying' it. The problem with SoundCloud is also that they also have separate Top 50's and New & Hot's for different genres, which we don't have implemented yet though.

theScrabi replied:
Well I don't like it ether. Because many services might make thins different form other services, even tho they might almost do "similar" things. Also I'm 100% sure people who will impelment their own NewPipe fronted will get that wrong. In my opinion It's the isiest solution to just crawl the translation. If it's not available in the language you selected then the content of the service will bin in the fall-back language anyway, so taking the effort ant translating only the headline would not be worth the work.

theScrabi replied:
I've add [KioskTranslator](https://github.com/TeamNewPipe/NewPipe/blob/f7534b3a0f72ca86f195de2a17e1bbab4669a74b/app/src/main/java/org/schabi/newpipe/util/KioskTranslator.java) with which local translation is possible.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/ExtractorHelper.java:
```diff
@@ -116,11 +116,29 @@ public PlaylistInfo call() throws Exception {
+        }));
+    }
+
+    public static Single<NextItemsResult> getMoreKisokItems(final int serviceId, final String url, final String nextStreamsUrl) {
```
thanks :)

wb9688 replied:
You could apply [this](https://gist.github.com/wb9688/dc1ac3f7154a3f14e970380d429b24e4) with `git am --signoff < kisok.patch` I guess.

theScrabi replied:
done

wb9688 commented:
Btw can you fix the merge conflicts?

theScrabi commented:
Resolved conflict.

wb9688 commented:
@theScrabi: Btw you haven't implemented the UI for eg. switching between Top 50 and New & Hot in SoundCloud's case, right?

theScrabi commented:
Not yet, atm its only trending.
Im willing to add filter, and selection at another time, and then not only for kiosk but also for channel and search, so you can also select latest first and such fuu.

theScrabi commented:
Not yet, atm its only trending.
Im willing to add filter, and selection at another time, and then not only for kiosk but also for channel and search, so you can also select latest first and such fuu.

theScrabi commented:
Looks like this now, and works ... somehow:
![screenshot_1506337606](https://user-images.githubusercontent.com/1891273/30805630-801bb6e4-a1f2-11e7-8230-95353865bbda.png)

wb9688 commented:
@theScrabi: I really don't like the idea of crawling the name
(@mauriciocolli: What do you think about that?). Besides that, is there a
reason why you're using that weird header instead of just putting the name
in the app bar as a page filter (so you could also choose for the other
kiosk types)? (And why those stars though?)

2017-09-25 13:07 GMT+02:00 Christian Schabesberger <notifications@github.com
>:

> Looks like this now, and works ... somehow:
> [image: screenshot_1506337606]
> <https://user-images.githubusercontent.com/1891273/30805630-801bb6e4-a1f2-11e7-8230-95353865bbda.png>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/TeamNewPipe/NewPipe/pull/732#issuecomment-331848944>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJU76ZKmj4hMCfglzn4E-sS6C6T3-8T8ks5sl4l-gaJpZM4PhlyH>
> .
>

karyogamy commented:
Since trending and other kiosk listings are network calls, wouldn't it be better to keep them as separate fragments? This way we can have a list of kiosk items on the main page (similar to "what's new" header on the subscription pager).

theScrabi commented:
@wb9688 the stars are just decoration. I thought it would look better. Tho i don't have an Idea how to style up the header.

However I thought about letting the user decide what he wants yo see on his startpage, so instead of just letting him see a kiosk we could add something to the settings so he can set a channel/the channel view/feed/blank page/and kiosk.

Now i would like to list the available kiosks in the dropdownmenue. (Not as separate fragments).

Also as headline for the startpage id like to put the service. Which was youtube in this case.

So in order to know what time of kiosk you are seeing id like to leave a header. Although its true the current header is ugly.

wb9688 commented:
>Also as headline for the startpage id like to put the service. Which was
youtube in this case.

@theScrabi: Why not your color idea instead?

mauriciocolli commented:
> Off-topic: is there a way to test this without an actual device and without Android SDK's emulator that never works for me?

Did you tried GenyMotion? (although atm, the official emulator is superior IMO)

> I really don't like the idea of crawling the name

I also disagree with this, and as I gave an example before, SoundCloud doesn't have all the languages that we support.

>  If it's not available in the language you selected then the content of the service will bin in the fall-back language anyway, so taking the effort ant translating only the headline would not be worth the work.

How does it not worth the effort? Everyone understand what "Trending/New & Hot" means?

--- 

In general, I think this PR is just too much early...

theScrabi commented:
> In general, I think this PR is just too much early...

I opened the PR also for discussion even tho we don't merge now.

> @theScrabi: Why not your color idea instead?

Both


So I thought it was better for the frontage to also be able to select other fragments as well. So I created a fragment selector in the settings:

![screenshot_1506442342](https://user-images.githubusercontent.com/1891273/30871581-a18c96cc-a2e7-11e7-9f9e-7a4574670497.png) ![screenshot_1506442349](https://user-images.githubusercontent.com/1891273/30871608-b348638c-a2e7-11e7-9123-9d9385450ebd.png) ![screenshot_1506442331](https://user-images.githubusercontent.com/1891273/30871578-9f08c9ac-a2e7-11e7-9739-4a52b3f3c319.png) ![screenshot_1506442477](https://user-images.githubusercontent.com/1891273/30871613-ba5c8c66-a2e7-11e7-9626-5b6b59c26248.png)

Also by far not finished.

timofonic commented:
@wb9688  @mauriciocolli About the Android emulator: Are you using Linux? If yes... What about using [Anbox](/anbox/anbox)?

wb9688 commented:
@timofonic: I'm using Debian Unstable, but Anbox tells me it doesn't
support Debian.

theScrabi commented:
I took care about some critics.
First of all I've added the [KioskTranslator](https://github.com/TeamNewPipe/NewPipe/blob/f7534b3a0f72ca86f195de2a17e1bbab4669a74b/app/src/main/java/org/schabi/newpipe/util/KioskTranslator.java) utill with which we can use resource to translate Kiosk name. Crawling the translation will still remain for compatibility reasons, tho it will not be in use in the android front end.

Secondary I've added a sub menu into the main menu, from which one can select the kiosks of the current services.

I am done with my work, so what do you guys think?

![screenshot_20171009-143234](https://user-images.githubusercontent.com/1891273/31338727-b653ba90-ad00-11e7-952e-72431e804df7.png)
My latest changes add a menu to the settings with which one can select Kiosks of multiple services. (It is disabled for now)

theScrabi commented:
I need your opinion.

KeizerDev commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -8,21 +9,43 @@
+    public int currentServiceId = -1;
+
+    /*//////////////////////////////////////////////////////////////////////////
+    // Konst
```
It should be const. Please no german here :stuck_out_tongue:

TobiGr commented:
I have very little time , so I can't say anything about the code at the moment. But after a short test, I can say: it works great! 👍 
I only found these small things:
1) The title of a kiosk is always shown twice - in my opinion once too much. I'd keep the one in the toolbar.
2) If you change the content of the main page in the settings, the change will only be applied after a restart of the app.
3) The margin when selecting the channel is to big and the padding is a little small, which makes the whole dialogue look squashed. Same things happen with the "Select a Kiosk" dialog.

<table><tr><td><img src="https://user-images.githubusercontent.com/17365767/31448380-93d883e2-aea4-11e7-9083-64398aa186bb.png"/></td>
<td><img src="https://user-images.githubusercontent.com/17365767/31448381-93f16ea2-aea4-11e7-99d6-0ca76500fd43.png"/>
</td></tr></table>
With tablet size, the channel selection does not look optimal either.

![screenshot_2017-10-11-16-36-28](https://user-images.githubusercontent.com/17365767/31448379-93bc4a1a-aea4-11e7-868f-bb124302f616.png)

PS: I am really looking forward to see this merged because I like the new features a lot!

theScrabi commented:
I removed the header, tho I have no idea how to make the UI update immediately.

karyogamy commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -112,6 +148,14 @@ public void onTabReselected(TabLayout.Tab tab) {
         public Fragment getItem(int position) {
             switch (position) {
+                case 0:
+                    if(PreferenceManager.getDefaultSharedPreferences(getActivity())
```
Why not just remove subscription page as a kiosk candidate? This kind of check might be problematic when there are more than 2 pagers on the main fragment.

theScrabi replied:
Well fix that once more pages are there. Also some people requested the Subscriptionpage to be shown directly when NewPipe is opened.

karyogamy commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -126,7 +170,89 @@ public CharSequence getPageTitle(int position) {
+        try {
+            SharedPreferences preferences =
+                    PreferenceManager.getDefaultSharedPreferences(getActivity());
+            final String set_main_page = preferences.getString(getString(R.string.main_page_content_key),
```
`set_main_page` should be in camel case.

theScrabi replied:
fixed

karyogamy commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -126,7 +170,89 @@ public CharSequence getPageTitle(int position) {
+    // Main page content
+    //////////////////////////////////////////////////////////////////////////*/
+
+    private Fragment getMainPageFramgent() {
```
Must we hard-code the strings in this method? If so, we should extract them.

theScrabi replied:
No, we discused about that. We came to conclution to not do that.

karyogamy replied:
I meant extracting the hardcoded strings and urls to a static variable or maybe put them in the Constants file.

theScrabi replied:
Ow I see.

theScrabi replied:
I created some static variables in the header instead

theScrabi replied:
fixed

karyogamy commented on app/src/main/java/org/schabi/newpipe/util/ServiceIconMapper.java:
```diff
@@ -0,0 +1,35 @@
+ * along with NewPipe.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+public class ServiceIconMapper {
```
Are we allowed to use these icons? I thought they are subject to copyright.

theScrabi replied:
Acording to @TheAssassin no, so I deaktivate it for now.
We later have to discuse that.

karyogamy commented on app/src/main/res/values/strings.xml:
```diff
@@ -266,4 +266,28 @@
+    <string name="select_a_kiosk">Select a kiosk</string>
+
+    <!-- Kiosk Names -->
+    <string name="kiosk">Kisok</string>
```
Spelling.

theScrabi replied:
fixed

coffeemakr commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -126,7 +170,89 @@ public CharSequence getPageTitle(int position) {
+        StreamingService service = NewPipe.getService(currentServiceId);
+        KioskList kl = service.getKioskList();
+        int i = 0;
+        for(final String ks : kl.getAvailableKisoks()) {
```
Typo in extractor

theScrabi replied:
fixed

coffeemakr commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java:
```diff
@@ -1,12 +1,168 @@
+                                            .putInt(getString(R.string.main_page_selected_service), service_id).apply();
+                                    defaultPreferences.edit()
+                                            .putString(getString(R.string.main_page_selectd_kiosk_id), kioskId).apply();
+                                    String summary = "";
```
Hard-coded string format. A better option would be to use string resource with the value 
`%1$s/%2$s`

theScrabi replied:
done

theScrabi commented:
I acidently merged "dev" into this branch, now I don't know how to squash the commits again :/
